core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -813,7 +813,14 @@ private[spark] class BlockManager(
               false
           }
         } else {
-          memoryStore.putBytes(blockId, size, level.memoryMode, () => bytes)
+          val memoryMode = level.memoryMode
+          memoryStore.putBytes(blockId, size, memoryMode, () => {
+            if (memoryMode == MemoryMode.OFF_HEAP) {
```

Comment:
No. I meant do we actually need to defensive copy here? All usage of `putBytes` across Spark have duplicated the byte buffer before passing it in. Is there any missing case we should do this defensive copy?